README.md
# zabbix-isp-speedtest
Zabbix template for monitoring ISP speeds using speedtest.net. 

To avoid timeouts when executing the speed test this template uses a combination of a trigger item and a trap item to asynchronously execute the speed test.

Latest version is v2.0.5 (2022-10-12). [Change Log](CHANGELOG.md)

## Limitations
* This template uses [External Checks](https://www.zabbix.com/documentation/2.2/manual/config/items/itemtypes/external) which are only supported on the Zabbix Server or a Proxy.

## Dependencies
* [speedtest-cli](https://github.com/sivel/speedtest-cli)
* [jq](https://github.com/stedolan/jq)
* [zabbix_sender](https://www.zabbix.com/documentation/current/manual/concepts/sender)
* atd job scheduler

## Installation (on a Zabbix Server)
* Install dependencies
* Copy isp-speedtest-trigger-atjob.sh and isp-speedtest-sender.sh to your Zabbix ExternalScripts location (default: /usr/lib/zabbix/externalscripts)
  * Ensure both scripts are executable by the Zabbix Agent user
* Copy userparameter_isp_speedtest.conf to the Zabbix Agent config folder (default: /etc/zabbix/zabbix_agentd.d)
  * Restart Zabbix Agent Service
* Import the template file into the Zabbix UI
* Link the template to the host

## Configuration

**Sender Macros**

The agent running the speedtest uses the zabbix sender to post the results back to the Zabbix server.

* **SPEEDTEST_SENDER_SERVER** (default: localhost): The hostname or IP address of the Zabbix server to send the results to

* **SPEEDTEST_SENDER_PORT** (default: 10051): The port of the Zabbix server

* **SPEEDTEST_SENDER_HOST** (default: Zabbix server): The host that the template is applied to so the sender can find the results item.

**Speedtest Macros**

Configure the execution of the speedtest runs

* **SPEEDTEST_UPDATE_INTERVAL** (default: 3h): How often to run the speedtest

* **SPEEDTEST_SERVER_ID** (default: blank): If filled in it will lock speedtest to a specific server (speedtest --list for available servers)

**Trigger Macros**

Configure trigger values for alerts

* **SPEEDTEST_TRIGGER_AVG_COUNT** (default: 3): Average values evaluated by triggers to prevent one bad run causing alerts

* **SPEEDTEST_TRIGGER_DOWNLOAD_LOW** (default: 100Mbps): Trigger when average download speeds fall below

* **SPEEDTEST_TRIGGER_UPLOAD_LOW** (default: 8Mbps): Trigger when average upload speeds fall below

* **SPEEDTEST_TRIGGER_LATENCY_HIGH** (default: 50ms): Trigger when average latency is above
  
## Template Items
* Speedtest Trigger (calls isp-speedtest-trigger-atjob script)
* Speedtest Results (trapper item - receives data from isp-speedtest-sender script)
  * Speedtest Download
  * Speedtest Upload
  * Speedtest Host (track which speedtest server gets used)
  * Speedtest Latency (rough latency to host)
  * Speedtest Success (boolean value if execution succeeded)
  * Speedtest Message (error message if any)

## Template Triggers
* Speedtest FAILED
* Download Speed LOW
* Upload Speed LOW
* Latency HIGH

## Template Graphs
* Download speed
* Upload speed
* Latency
* Summary graph with download, upload, and latency

The 3 individual graphs are included on a monitoring screen as well


CHANGELOG.md
# zabbix-speedtest.net change log

## v2.0.5 - 2022-10-12
* Fix [Issue 7](https://github.com/robmccready/zabbix-isp-speedtest/issues/7)
* Fix [Issue 8](https://github.com/robmccready/zabbix-isp-speedtest/issues/8)

## v2.0.4 - 2021-04-12
* Escape quotes in error message passed back to zabbix. Handles odd issue with recent (now fixed) bug in speedtest-cli [Handle case where ignoreids is empty or contains empty ids](https://github.com/sivel/speedtest-cli/commit/cadc68b5aef20f28648072cf07a8f155639b81dd)

## v2.0.3 - 2020-11-11
* Rename All-In-One report to "Speedtest Summary"
 * Update graph colors and summary graph to be cleaner
* Add high latency trigger (default 50ms)
* Increase Trend Histories to 750d to allow year on year comparison
* Add macros for configuring zabbix server/port for sender


## v2.0.0 - 2020-10-30
Scripts refactored to leverage atd and zabbix_sender to disconnect the execution of the speedtest from the zabbix maximum timeout of 30 seconds.[Escaping timeouts with atd](https://zabbix.org/wiki/Escaping_timeouts_with_atd)
* Add trigger script isp-speedtest-trigger-atjob.sh
* Refactor (and rename) speedtest script to use zabbix_sender: isp-speedtest-sender.sh
* Add support for specified speedtest server ID (speedtest --server <ID>)
* Add all-in-one report with download, upload, and latency
* Reskinned graphs to use speedtest.net download (greenish) and upload (purplish) colors


## v1.1.0 - 2019-09-18
* Convert from External Check to Zabbix Agent execution
* Rename items to net.isp.speedtest prefix
* Rename some files

## v1.0.0 - 2019-09-17
* Updated item names and descriptions
* Made execution script handle missing dependencies, execution errors, parsing results and returning a simplified JSON object which includes error info (if any)



test_send.sh
#!/bin/bash
./isp-speedtest-sender.sh localhost 10051 "Zabbix server"



test_trigger.sh
#!/bin/bash
./isp-speedtest-trigger-atjob.sh localhost 10051 "Zabbix server"



speedtest.sh
#!/bin/sh
#
# Calls speedtest cli command and parses result to a more readable format
# {
#   "success": 1 (0 or 1)
#   "message": "No problems"
#   "host": "some.speedtest.server.com",
#   "latency": 123.4,
#   "download": 1234567.890,
#   "upload": 1234567.890
#  }
#


#
# Validate dependencies are installed
#
if ! [ -x "$(command -v speedtest)" ]; then
  echo '{"success": 0, "message": "speedtest command not found"}'
  exit 1
fi

if ! [ -x "$(command -v jq)" ]; then
  echo '{"success": 0, "message": "jq command not found"}'
  exit 1
fi

#
# Perform speedtest and check exit code
#
SPEEDTEST_OUTPUT=$(speedtest --json)
retVal=$?
if [ $retVal -ne 0 ]; then
  echo '{"success": 0, "message": "speedtest failure code: '$retVal'"}'
  exit 1
fi

#
# Decode output json
#
host=$(echo $SPEEDTEST_OUTPUT | jq -r '.server.host')
latency=$(echo $SPEEDTEST_OUTPUT | jq -r '.server.latency')
download=$(echo $SPEEDTEST_OUTPUT | jq -r '.download')
upload=$(echo $SPEEDTEST_OUTPUT | jq -r '.upload')

#
# Validate result fields
#
if [ -z "$host" ]; then
  echo '{"success": 0, "message": "Host not found in result"}'
  exit 1
fi

if [ -z "$latency" ]; then
  echo '{"success": 0, "message": "Latency not found in result"}'
  exit 1
fi

if [ -z "$download" ]; then
  echo '{"success": 0, "message": "Download not found in result"}'
  exit 1
fi

if [ -z "$upload" ]; then
  echo '{"success": 0, "message": "Upload not found in result"}'
  exit 1
fi


#
# Success
#
echo '{"success": 1, "message": "no problem", "host": "'$host'", "latency": '$latency', "download": '$download', "upload": '$upload'}'


isp-speedtest-sender.sh
#!/bin/bash
#
# Calls speedtest cli command and parses result to a more readable format
# {
#   "success": 1 (0 or 1)
#   "message": "No problems"
#   "host": "some.speedtest.server.com",
#   "latency": 123.4,
#   "download": 1234567.890,
#   "upload": 1234567.890
#  }
#

if [ "$#" -lt 3 ]; then
	echo "isp-speedtest-sender.sh requires minimum 3 arguments (server, port, host) speedtest server ID is an optional 4th argument"
	exit 1
fi

zabbix_server=$1
zabbix_port=$2
zabbix_host=$3
speedtest_server_id=$4

function zabbix_send {
	echo $1
	zabbix_sender --zabbix-server "${zabbix_server}" --port ${zabbix_port} --host "${zabbix_host}" --key net.isp.speedtest.results --value "$1" -vv
}

#
# Validate dependencies are installed
#
if ! [ -x "$(command -v zabbix_sender)" ]; then
  echo "zabbix_sender command not found"
  exit 1
fi

#
# Other dependency failures can be reported back to zabbix via the sender
#

if ! [ -x "$(command -v speedtest)" ]; then
  zabbix_send '{"success": 0, "message": "speedtest command not found"}'
  exit 1
fi

if ! [ -x "$(command -v jq)" ]; then
  zabbix_send '{"success": 0, "message": "jq command not found"}'
  exit 1
fi

#
# Perform speedtest and check exit code
#
if [ -z "$speedtest_server_id" ]; then
	SPEEDTEST_OUTPUT=$(speedtest --secure --json 2>&1)
else
	SPEEDTEST_OUTPUT=$(speedtest --secure --json --server ${speedtest_server_id} 2>&1)
fi

retVal=$?
if [ $retVal -ne 0 ]; then
  SPEEDTEST_OUTPUT=$(echo $SPEEDTEST_OUTPUT | sed 's/\"//g')
  zabbix_send "{\"success\": 0, \"message\": \"speedtest failure code [$retVal], output: [${SPEEDTEST_OUTPUT}]\", \"latency\": 0, \"download\": 0, \"upload\": 0, \"host\": null}"
  exit 1
fi

#
# Decode output json
#
echo $SPEEDTEST_OUTPUT
host=$(echo $SPEEDTEST_OUTPUT | jq -r '.server.host')
host_id=$(echo $SPEEDTEST_OUTPUT | jq -r '.server.id')
latency=$(echo $SPEEDTEST_OUTPUT | jq -r '.server.latency')
download=$(echo $SPEEDTEST_OUTPUT | jq -r '.download')
upload=$(echo $SPEEDTEST_OUTPUT | jq -r '.upload')

#
# Validate result fields
#
if [ -z "$host" ]; then
  zabbix_send '{"success": 0, "message": "Host not found in result"}'
  exit 1
fi

if [ -z "$latency" ]; then
  zabbix_send '{"success": 0, "message": "Latency not found in result"}'
  exit 1
fi

if [ -z "$download" ]; then
  zabbix_send '{"success": 0, "message": "Download not found in result"}'
  exit 1
fi

if [ -z "$upload" ]; then
  zabbix_send '{"success": 0, "message": "Upload not found in result"}'
  exit 1
fi


#
# Success
#
zabbix_send "{\"success\": 1, \"message\": \"no problem\", \"host\": \"$host\", \"server_id\": \"$host_id\", \"latency\": $latency, \"download\": $download, \"upload\": $upload}"
